### PR TITLE
Add support for partial checkout filters

### PR DESCRIPTION
**Pull request by finleygn**
A change to add support for sparse checkouts has been added recently (#1369). This extends upon that to allow support for other filters, and the ability to filter without a sparse checkout.

```yaml
- uses: actions/checkout@v3
  with:
    filter: tree:0
```

Inspired by this blog post -> https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone

### CODE REVIEW

**Review by Spiritaine — APPROVED** [review]

**Comment by gabrielyotoo** [comment]
bump +1 this would be nice to use with Yarn

**Comment by petergaultney** [comment]
i've used this fork/branch successfully with both `tree:0` and `blob:none`. Thank you!

**Comment by flying-sheep** [comment]
I think `filter: 'tree:0'` would be a much better default than `fetch-depth: 1`.

It would be almost exactly as small, but without breaking the Git history.

File: __test__/verify-fetch-filter.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  echo "Expected ./fetch-filter/.git folder to exist"
+  exit 1
+fi
+
+# Verify .git/config contains partialclonefilter
+
+CLONE_FILTER=$(git config --local --get remote.origin.partialclonefilter)
+
+if [ "$CLONE_FILTER" != "blob:none" ]; then
+  echo "Expected ./fetch-filter/.git/config to have 'remote.origin.partialclonefilter' set to 'blob:none'"
```
Comment (gabrielyotoo) [inline __test__/verify-fetch-filter.sh:14]: did you miss exit 1 here?

Comment (finleygn) [reply __test__/verify-fetch-filter.sh:14]: yep, good spot 😳

**Comment by kalanzai** [comment]
This is exactly what we need in our repo. Speeds up checkout by a factor of 4 without breaking git history so our existing logic that uses `git diff` and `git blame` will still work ❤️

**Comment by cory-miller** [comment]
Thanks for your contribution.

Do you mind adding an update to https://github.com/actions/checkout/blob/72f2cec99f417b1a1c5e2e88945068983b7965f9/action.yml#L56C19-L60 with the new setting with a note that it will override the `sparseCheckout` option if also set?

The rest LGTM.

**Review by Spiritaine — APPROVED** [review]

**Comment by finleygn** [comment]
Will take a look as to whats causing the test to fail soon...

**Comment by finleygn** [comment]
@cory-miller the test actions *should* pass now 🙏

**Review by cory-miller — APPROVED** [review]

**Comment by cory-miller** [comment]
Thanks for working on this!

I will build a release soon.

**Comment by cory-miller** [comment]
`actions/checkout@v4` or `actions/checkout@v4.1.0` are updated with this change.

**Comment by Rycieos** [comment]
@finleygn thanks for working on this!

@cory-miller the issues #172, #663, and #1152 should likely be closed.

**Comment by benoit-pierre** [comment]
This does not seem to apply to submodules.

**Comment by Kurt-von-Laven** [comment]
> I think `filter: 'tree:0'` would be a much better default than `fetch-depth: 1`.
> 
> It would be almost exactly as small, but without breaking the Git history.

This is a cool feature, but please don't make the default checkout larger for everyone since most users don't need the Git history in their CI pipelines. We do often need the history, so we tried switching from `fetch-depth: 0` to `filter: tree:0`, and our CI pipelines became twice as slow _overall_. Then we tried `filter: blob:none`, and that still slowed down our CI pipelines by 50%. I suspect this is happening because some tool (e.g., Gitleaks) is causing Git to fetch the missing data piecemeal rather than more efficiently all at once by attempting to access trees or blobs that aren't available locally. I don't mean this as a knock on partial clones; everyone should always measure the impact of performance changes in their own specific use case. I would guess that treeless clones are a net performance benefit if and only if no trees or blobs from previous commits are needed, because GitHub has already put so much effort into optimizing checkout times in GitHub Actions. Similarly, I would guess that blobless clones are a net performance benefit if and only if no blobs from previous commits are needed.

**Comment by jakebailey** [comment]
I think there's something missing from the docs here; you still need to explicitly set `fetch-depth: 0` in order to get all of history, making a full workflow look like:

```yml
- use: actions/checkout@v4
  with:
    fetch-depth: 0
    filter: blob:none
```

Without setting it, you still only get the most recent commit.

For the TypeScript repo:

- `fetch-depth: null`, `filter: null`: 12s, but only one commit.
- `fetch-depth: null`, `filter: blob:none`: 30s, but still only one commit!
- `fetch-depth: 0`, `filter: null`, 4m6s: full history
- `fetch-depth: 0`, `filter: filter:none`, 1m37s: full history

**Comment by Kurt-von-Laven** [comment]
@jakebailey, did you mean `filter: blob:none` rather than `filter: filter:none` for the last bullet?

**Comment by jakebailey** [comment]
Yes, that's just a typo on my part.

**Pull request merged**